core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java

```diff
@@ -112,29 +113,47 @@ private boolean shouldTrigger(Run upstreamBuild, TaskListener listener) {
         if (job == null) {
             return false;
         }
+
+        boolean downstreamVisible = false;
+        boolean downstreamDiscoverable = false;
+
         // This checks Item.READ also on parent folders; note we are checking as the upstream auth currently:
-        boolean downstreamVisible = jenkins.getItemByFullName(job.getFullName()) == job;
+        try {
+            downstreamVisible = jenkins.getItemByFullName(job.getFullName()) == job;
+        } catch (AccessDeniedException ex) {
+            // Fails because of missing Item.READ but upstream user has Item.DISCOVER
+            downstreamDiscoverable = true;
+        }
+
         Authentication originalAuth = Jenkins.getAuthentication();
         Job upstream = upstreamBuild.getParent();
         Authentication auth = Tasks.getAuthenticationOf((Queue.Task) job);
         if (auth.equals(ACL.SYSTEM) && !QueueItemAuthenticatorConfiguration.get().getAuthenticators().isEmpty()) {
             auth = Jenkins.ANONYMOUS; // cf. BuildTrigger
         }
+
         SecurityContext orig = ACL.impersonate(auth);
+        Item authUpstream = null;
         try {
-            if (jenkins.getItemByFullName(upstream.getFullName()) != upstream) {
-                if (downstreamVisible) {
-                    // TODO ModelHyperlink
-                    listener.getLogger().println(Messages.ReverseBuildTrigger_running_as_cannot_even_see_for_trigger_f(auth.getName(), upstream.getFullName(), job.getFullName()));
-                } else {
-                    LOGGER.log(Level.WARNING, "Running as {0} cannot even see {1} for trigger from {2} (but cannot tell {3} that)", new Object[] {auth.getName(), upstream, job, originalAuth.getName()});
-                }
-                return false;
-            }
+            authUpstream = jenkins.getItemByFullName(upstream.getFullName());
             // No need to check Item.BUILD on downstream, because the downstream project’s configurer has asked for this.
+        } catch (AccessDeniedException ade) {
+            // Fails because of missing Item.READ but downstream user has Item.DISCOVER
         } finally {
             SecurityContextHolder.setContext(orig);
         }
+
+        if(authUpstream != upstream) {
+            if (downstreamVisible) {
+                // TODO ModelHyperlink
+                listener.getLogger().println(Messages.ReverseBuildTrigger_running_as_cannot_even_see_for_trigger_f(auth.getName(),
+                        upstream.getFullName(), job.getFullName()));
+            } else  {
+                LOGGER.log(Level.WARNING, Messages.ReverseBuildTrigger_running_as_cannot_even_see_but_cannot_tell(auth.getName(),
+                        downstreamDiscoverable ? "READ" : "DISCOVER", upstream, job, originalAuth.getName()));
```

Comment:
I was also confusing as well. Here is my understanding.

Bob runs upstream. Let's call Bob "upstream user"
Alice runs downstream. Let's call Alice "downstream user"

![jenkins-42707-summary](https://cloud.githubusercontent.com/assets/6858329/25221955/4becf4ba-25fa-11e7-93f8-c417890955f3.png)

* _trigger_: should trigger
* _message_: display a message in Upstream console
* _log_: display a message in Jenkins log
* Font in Red: The actual issue JENKINS-42707 (we see the exception in the console)
* Circle in Blue: I'm not sure what is expected.

Thanks for helping.